### DISCUSSION TITLE
Inconsistent State

### DISCUSSION THREAD

**Discussion by ghost** [pmndrs/zustand#1556] [Q&A]
## Summary

Hello everyone,
i have a problem with a inconsistent state.

My Store:
```
    movementDialog: {
        source: "",
        draggableId: "",
        destination: ""
    },
        updateMovementDialog: (newState: Dialog) => set(() => ({
        movementDialog: {...newState}
    })),
```

My Logic:

[get the data]

const newDialog: Dialog = {
            destination: result.destination,
            source: result.source,
            draggableId: result.draggableId
        }
        updateMovementDialog(newDialog);

        if (!movementDialog.destination) {
            console.log("movementdialog.destination");
            return;
        }`

1. The function "MyLogic accepts a result as input. Afterwards i like to save the state in zustand (MyStore).
That are different files of course. I copied the relevant parts in here.
2. I save the state via updateMovementDialog
3. Than i like i load the state over movementDialog variable, but in the variable is the "OLD STATE namely:

movementDialog: {
        source: "",
        draggableId: "",
        destination: ""
    },
".

So the state is not changed, inconsistent and was not muted, if i call the function "my logic" a second time, 
than of course is the data saved consistent. 
So it needs time, so save the state in zustand.
What would be a good solution for this issue, it is not possible to call the function twice.

Thanks for your help.

**Comment by dai-shi** [comment]
Can you please try to create a very small reproduction with https://codesandbox.io/s/react-new ?

**Comment by ghost** [comment]
https://codesandbox.io/s/optimistic-cherry-sh5slc?file=/src/App.tsx

**Reply by dai-shi** [reply]
This is so-called stale value behavior with react hooks.
Nothing is wrong, and the same behavior can be observed with `useState`:
https://codesandbox.io/s/priceless-wind-o6grre

Thanks to zustand, you can read your store in your callback function.
https://codesandbox.io/s/angry-noether-czbvn7

**Comment by ghost** [comment]
Yes, but than i have the problem, i have to use that everywhere in the code afterwards.
useStorage().getState() and so on and than access to the state
and not via movementDialog. Namely until i the state saved.

Is there no possibiltiy to wait until the hook is finsihed?

**Reply by dai-shi** [reply]
It's not about waiting something.

It's how JavaScript works.
```js
let count = 0;

const fn = () => {
  const currentCount = count;
  const handleClick = () => {
    console.log(currentCount);
    count += 1;
    console.log(currentCount);
  };
  setTimeout(handleClick, 500);
};

fn();
```

**Reply by ghost** [reply]
Yes. I know. I would like to wait until the state is saved. 
Afterwards i like to read it out from zustand.

**Reply by ghost** [reply]
because now i have to use everywhere
useStorage().getState()
